javalib/src/main/scala/java/lang/Throwables.scala

```diff
@@ -49,49 +49,126 @@ private[lang] object StackTrace {
   }
 }
 
-class Throwable(s: String, private var e: Throwable)
+class Throwable protected (s: String,
+                           private[this] var e: Throwable,
+                           enableSuppression: scala.Boolean,
+                           writableStackTrace: scala.Boolean)
     extends Object
     with java.io.Serializable {
+
+  def this(message: String, cause: Throwable) =
+    this(message, cause, true, true)
+
   def this() = this(null, null)
+
   def this(s: String) = this(s, null)
+
   def this(e: Throwable) = this(if (e == null) null else e.toString, e)
 
-  private var stackTrace: Array[StackTraceElement] = _
+  private[this] var stackTrace: Array[StackTraceElement] = _
 
-  fillInStackTrace()
+  if (writableStackTrace)
+    fillInStackTrace()
 
-  def initCause(cause: Throwable): Throwable = {
-    e = cause
-    this
-  }
+  // We use an Array rather than, say, a List, so that Throwable does not
+  // depend on the Scala collections.
+  private[this] var suppressed: Array[Throwable] = _
 
-  def getMessage(): String = s
+  final def addSuppressed(exception: Throwable): Unit = {
 
-  def getCause(): Throwable = e
+    if (exception == null) {
+      // Yes, JVM message uses 'cannot' (sic) & has terminal period/full_stop!
+      throw new java.lang.NullPointerException(
+        "Cannot suppress a null exception.")
+    }
 
-  def getLocalizedMessage(): String = getMessage()
+    if (exception == this) {
+      throw new java.lang.IllegalArgumentException(
+        "Self-suppression not permitted")
+    }
+
+    if (enableSuppression) this.synchronized {
+      if (suppressed eq null) {
+        suppressed = Array(exception)
+      } else {
+        val length        = suppressed.length
+        val newSuppressed = new Array[Throwable](length + 1)
+        System.arraycopy(suppressed, 0, newSuppressed, 0, length)
+        newSuppressed(length) = exception
+        suppressed = newSuppressed
+      }
+    }
+  }
 
   def fillInStackTrace(): Throwable = {
-    this.stackTrace = StackTrace.currentStackTrace()
+    // currentStackTrace should be handling exclusion in its own
+    // critical section, but does not. So do
+    if (writableStackTrace) this.synchronized {
+      this.stackTrace = StackTrace.currentStackTrace()
```

Comment:
check `this.stackTrace ne null`?

Reply:
@hepin1989 Thank you for the review & suggestion. Always good to have another
pair of eyes go over the code. 

I concur that `null.synchronized` and `null.stackTrace` would not be a good thing to happen.

I believe that if the code in lines 106 & 107 are executing then `this` is an
instantiated object and can not be null.  Am I missing something?

Reply:
The code is correct as is.

Reply:
@LeeTibbert 
```scala
if(this.stackTrace ne null) this 
```
is not needed?

Reply:
No, it isn't, and would even go against what the JVM does. If you first call `setStackTrace()` then `fillInStackTrace()`, the latter will overwrite the former.